sui-execution/latest/sui-move-natives/src/crypto/vdf.rs

```diff
@@ -0,0 +1,182 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+use crate::object_runtime::ObjectRuntime;
+use crate::NativesCostTable;
+use fastcrypto_vdf::class_group::discriminant::Discriminant;
+use fastcrypto_vdf::class_group::QuadraticForm;
+use fastcrypto_vdf::vdf::wesolowski::StrongVDF;
+use fastcrypto_vdf::vdf::wesolowski::StrongVDFVerifier;
+use fastcrypto_vdf::ToBytes;
+use move_binary_format::errors::PartialVMResult;
+use move_core_types::gas_algebra::InternalGas;
+use move_core_types::vm_status::StatusCode;
+use move_vm_runtime::{native_charge_gas_early_exit, native_functions::NativeContext};
+use move_vm_types::natives::function::PartialVMError;
+use move_vm_types::{
+    loaded_data::runtime_types::Type,
+    natives::function::NativeResult,
+    pop_arg,
+    values::{Value, VectorRef},
+};
+use smallvec::smallvec;
+use std::collections::VecDeque;
+
+pub const INVALID_INPUT_ERROR: u64 = 0;
+pub const NOT_SUPPORTED_ERROR: u64 = 1;
+
+fn is_supported(context: &NativeContext) -> bool {
+    context
+        .extensions()
+        .get::<ObjectRuntime>()
+        .protocol_config
+        .enable_vdf()
+}
+
+#[derive(Clone)]
+pub struct VDFCostParams {
+    pub vdf_verify_cost: Option<InternalGas>,
+    pub hash_to_input_cost: Option<InternalGas>,
+}
+
+/***************************************************************************************************
+ * native fun vdf_verify_internal
+ *
+ * Implementation of the Move native function `vdf::verify_vdf_internal(
+ *      discriminant: &vector<u8>,
+ *      input: &vector<u8>,
+ *      output: &vector<u8>,
+ *      proof: &vector<u8>,
+ *      iterations: u64): bool`
+ *
+ * Gas cost: verify_vdf_cost
+ **************************************************************************************************/
+pub fn vdf_verify_internal(
+    context: &mut NativeContext,
+    ty_args: Vec<Type>,
+    mut args: VecDeque<Value>,
+) -> PartialVMResult<NativeResult> {
+    let cost = context.gas_used();
+    if !is_supported(context) {
```

Comment:
Is there an issue at all with the fact that we are not charging any gas for this call in the case where it's not supported? (i.e. should there be some nominal gas cost for doing this?)

Reply:
@amnn it's minimal, IMO we're good